dcos_test_utils/ssh_client.py

```diff
@@ -136,3 +132,133 @@ def wait_for_ssh_connection(self, host: str, port: int=22) -> None:
+
+    @asyncio.coroutine
+    def run_cmd_return_dict_async(self, cmd: list) -> dict:
+        """
+        Runs cmd with a pTTY and times out if necessary
+        Returns a dict of data about the process
+        """
+        with make_slave_pty() as slave_pty:
+            process = yield from asyncio.create_subprocess_exec(
+                *cmd, stdout=asyncio.subprocess.PIPE,
+                stderr=asyncio.subprocess.PIPE,
+                stdin=slave_pty,
+                env={'TERM': 'linux'})
+            stdout = b''
+            stderr = b''
+            try:
+                stdout, stderr = yield from asyncio.wait_for(process.communicate(), self.process_timeout)
+            except asyncio.TimeoutError:
+                try:
+                    process.terminate()
+                except ProcessLookupError:
+                    log.info('process with pid {} not found'.format(process.pid))
+                log.error('timeout of {} sec reached. PID {} killed'.format(self.process_timeout, process.pid))
+
+        return {
+            "cmd": cmd,
+            "stdout": stdout,
+            "stderr": stderr,
+            "returncode": process.returncode,
+            "pid": process.pid
+        }
+
+    @asyncio.coroutine
+    def run_async(self, host: str, cmd: list) -> dict:
+        """ runs cmd as arguments on host
+        """
+        hostname, port = parse_ip(host)
+        with self.tunnel(hostname, port) as t:
+            full_cmd = t.base_cmd + [t.target] + cmd
+            log.debug('executing command {}'.format(full_cmd))
+            result = yield from self.run_cmd_return_dict_async(full_cmd)
+        result['host'] = host
+        return result
+
+    @asyncio.coroutine
+    def copy_async(self, host: str, local_path: str, remote_path: str, recursive: bool) -> dict:
+        """ instructs SCP to copy local_path to remote_path on host. recursive copying may be enabled
+        """
+        hostname, port = parse_ip(host)
+        copy_command = []
+        if recursive:
+            copy_command.append('-r')
+        remote_full_path = '{}@{}:{}'.format(self.user, hostname, remote_path)
+        copy_command += [local_path, remote_full_path]
+        full_cmd = ['/usr/bin/scp'] + SHARED_SSH_OPTS + ['-P', str(port), '-i', self.key_path] + copy_command
+        log.debug('copy with command {}'.format(full_cmd))
+        result = yield from self.run_cmd_return_dict_async(full_cmd)
+        result['host'] = host
+        return result
+
+    @asyncio.coroutine
+    def run_command_on_hosts(self, coroutine_name: str, *args) -> list:
+        """ coroutine to run a single coroutine against all hosts in the MultiRunner
+        """
+        tasks = yield from self.start_command_on_hosts(coroutine_name, *args)
+        yield from asyncio.wait(tasks)
+        return [task.result() for task in tasks]
+
+    @asyncio.coroutine
+    def start_command_on_hosts(self, coroutine_name: str, *args, sem: asyncio.Semaphore=None) -> list:
```

Comment:
I doubt if carrying a semaphore instance as an argument is going to useful (or easy to reason with). num_of_concurrent_requests / parallelism as an argument be easier to reason.

Reply:
It is implemented for a specific use and is otherwise option, see here: https://github.com/dcos/dcos-launch/pull/75/files#diff-a0399dc178afa3d227eae419055f1239R220